File: apps/mobile/src/types/holding.ts
/**
 * 持仓领域类型
 */
import type { Currency } from './common';

/** 市场枚举（原型里支持 A 股 / 港股 / 美股等） */
export type Market = 'CN_A' | 'HK' | 'US' | 'SG' | 'OTHER';

/** 股票基础信息 */
export interface Stock {
  symbol: string;
  name: string;
  market: Market;
  currency: Currency;
  /** 最新价（由行情接口刷新） */
  lastPrice?: number;
  /** 当日涨跌幅（百分比） */
  dayChangePct?: number;
  /** 股息率（年化百分比） */
  dividendYield?: number;
}

/** 一条持仓记录 */
export interface Holding {
  id: string;
  userId: string;
  stock: Stock;
  /** 持有数量（股） */
  shares: number;
  /** 成本均价（stock.currency 计） */
  avgCost: number;
  /** 买入日期（ISO） */
  firstBuyAt?: string;
  /** 备注 */
  note?: string;
  createdAt: string;
  updatedAt: string;
}

/** 持仓组合概览 */
export interface PortfolioSummary {
  baseCurrency: Currency;
  /** 总市值（折算到 baseCurrency） */
  totalMarketValue: number;
  /** 总成本 */
  totalCost: number;
  /** 浮动盈亏 */
  unrealizedPnl: number;
  unrealizedPnlPct: number;
  /** 年化预期分红 */
  annualDividendEstimate: number;
}

File: apps/api/app/schemas/holding.py
"""
持仓相关 Pydantic schema
"""
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class HoldingBase(BaseModel):
    """持仓公共字段"""

    symbol: str = Field(..., max_length=32)
    shares: float = Field(..., ge=0)
    avg_cost: float = Field(..., ge=0)
    first_buy_at: datetime | None = None
    note: str | None = Field(default=None, max_length=512)


class HoldingCreate(HoldingBase):
    """创建持仓入参"""


class HoldingUpdate(HoldingBase):
    """更新持仓入参"""


class HoldingOut(HoldingBase):
    """持仓响应"""

    id: str
    user_id: str
    market: str
    currency: str
    name: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PortfolioSummary(BaseModel):
    """组合概览"""

    base_currency: str
    total_market_value: float
    total_cost: float
    unrealized_pnl: float
    unrealized_pnl_pct: float
    annual_dividend_estimate: float


class ScanHoldingsIn(BaseModel):
    """AI 扫描截图入参"""

    image: str = Field(..., description="base64 编码的图片")


class ScanHoldingsOut(BaseModel):
    """AI 扫描解析结果"""

    candidates: list[HoldingCreate]
